src/shallow/ShallowWrapper.lua

```diff
@@ -0,0 +1,202 @@
+			getChildren(subVirtualNode, results, maxDepth)
+		end
+	else
+		local childWrapper = ShallowWrapper.new(
+			virtualNode,
+			maxDepth
+		)
+
+		table.insert(results, childWrapper)
+	end
+end
+
+local function filterProps(props)
+	if props[Children] == nil then
+		return props
+	end
+
+	local filteredProps = {}
+
+	for key, value in pairs(props) do
+		if key ~= Children then
+			filteredProps[key] = value
+		end
+	end
+
+	return filteredProps
+end
+
+function ShallowWrapper.new(virtualNode, maxDepth)
+	virtualNode = findNextVirtualNode(virtualNode, maxDepth)
+
+	local wrapper = {
+		[InternalData] = {
+			virtualNode = virtualNode,
+			childrenMaxDepth = maxDepth - 1,
+			virtualNodeChildren = maxDepth == 0 and {} or virtualNode.children,
+			instance = virtualNode.hostObject,
+		},
+		type = getTypeFromVirtualNode(virtualNode),
+		props = filterProps(virtualNode.currentElement.props),
+		hostKey = virtualNode.hostKey,
+	}
+
+	return setmetatable(wrapper, ShallowWrapperMetatable)
+end
+
+function ShallowWrapperPublic:childrenCount()
+	local count = 0
+	local internalData = self[InternalData]
+
+	for _, virtualNode in pairs(internalData.virtualNodeChildren) do
+		local element = virtualNode.currentElement
+		count = count + countChildrenOfElement(element)
+	end
+
+	return count
+end
+
+function ShallowWrapperPublic:find(constraints)
+	VirtualNodeConstraints.validate(constraints)
+
+	local results = {}
+	local children = self:getChildren()
+
+	for i=1, #children do
+		local childWrapper = children[i]
+		local childInternalData = childWrapper[InternalData]
+
+		if VirtualNodeConstraints.satisfiesAll(childInternalData.virtualNode, constraints) then
+			table.insert(results, childWrapper)
+		end
+	end
+
+	return results
+end
+
+function ShallowWrapperPublic:findUnique(constraints)
+	local children = self:getChildren()
+
+	if constraints == nil then
+		assert(
+			#children == 1,
+			("expect to contain exactly one child, but found %d"):format(#children)
+		)
+		return children[1]
+	end
+
+	local constrainedChildren = self:find(constraints)
+
+	assert(
+		#constrainedChildren == 1,
+		("expect to find only one child, but found %d"):format(#constrainedChildren)
+	)
+
+	return constrainedChildren[1]
+end
+
+function ShallowWrapperPublic:getChildren()
+	local results = {}
+	local internalData = self[InternalData]
+
+	for _, childVirtualNode in pairs(internalData.virtualNodeChildren) do
+		getChildren(childVirtualNode, results, internalData.childrenMaxDepth)
+	end
+
+	return results
+end
+
+function ShallowWrapperPublic:getInstance()
+	local internalData = self[InternalData]
+
+	return internalData.instance
+end
+
+function ShallowWrapperPublic:matchSnapshot(identifier)
+	assert(typeof(identifier) == "string", "Snapshot identifier must be a string")
+
+	local snapshotResult = Snapshot.createMatcher(identifier, self)
+
+	snapshotResult:match()
+end
+
+function ShallowWrapperPublic:snapshotToString()
+	return Snapshot.toString(self)
```

Comment:
should we just have a __tostring meta method for this on ShallowWrapper?

Reply:
Hum I am not sure about `tostring(ShallowWrapper)`... It not very intuitive I think, because I would expect to receive a string version of a shallow wrapper instead of a generated snapshot from it. What do you think?

Reply:
I was trying to think of what to name this, similar to react-test-renderer toJSON(). Maybe toSnapshotString is a bit better name and ya reserve tostring(ShallowWrapper) for a future string representation that is more focused on the shallow's structure.